### PR TITLE
Take into account the data queued in the sender

### LINKED ISSUE #970: Extend bufferedAmount property
The bufferedAmount property is good for your application to control how much messages you can send (to avoid memory overhead), but it works only if you disable perMessageDeflate.

I propose to extend the property by calculating _sender.queue size as well.

Example:

    get bufferedAmount () {
        var amount = 0;
        var senderQueueSize = 0;
        
        if (this._socket) amount = this._socket.bufferSize || 0;
        
        if (this._sender && this._sender.queue) {
          for (i = 0; i < this._sender.queue.length; i += 1) {
            senderQueueSize += this._sender.queue[i].data.length;
          }
        }
        return amount + senderQueueSize;
    }

The best solution is to avoid cycle at all by adding queueSize variable to _sender and add/subtract length when adding a new element to queue.

    get bufferedAmount () {
        var amount = 0;
        var senderQueueSize = 0;
        
        if (this._socket) amount = this._socket.bufferSize || 0;
        return amount + this._sender.queueSize;
    }

### PR DESCRIPTION
**Pull request by lpinca**
This makes the `bufferedAmount` getter take into account the data queued in the sender.

Fixes #970.

cc: @alexions

### CODE REVIEW

File: lib/Sender.js [hunk 1]
```diff
@@ -86,10 +90,23 @@ class Sender {
...
+      if (data instanceof ArrayBuffer) {
+        data = Buffer.from(data);
+      } else if (ArrayBuffer.isView(data)) {
+        data = viewToBuffer(data);
+      } else {
+        data = Buffer.from(data);
+        readOnly = false;
+      }
+    }
+
```
Comment (alexions) [inline lib/Sender.js]: I'd move this block of code to a separate function, like 

    makeBuffer (data) {
      var readOnly = true;

      if (data && !Buffer.isBuffer(data)) {
        if (data instanceof ArrayBuffer) {
          data = Buffer.from(data);
        } else if (ArrayBuffer.isView(data)) {
          data = viewToBuffer(data);
        } else {
          data = Buffer.from(data);
          readOnly = false;
        }
      }

      return [data, readOnly];
    }

    Usage:
    [data, readOnly] = makeBuffer(data);

Comment (lpinca) [inline lib/Sender.js]: I also don't like having this code duplication. The reason for not using a separate function is to avoid creating an array.

Comment (alexions) [inline lib/Sender.js]: Agree. Array can make additional overhead... ok :) Speed is better than duplicates :)

Comment (lpinca) [inline lib/Sender.js]: It should be benchmarked to see if it actually makes any difference :)

Comment (lpinca) [inline lib/Sender.js]: Another option would be to add the `readOnly` flag to the returned buffer. For example:

```js
const kReadOnly = Symbol('read-only');

function toBuffer(data) {
  var readOnly = true;

  if (!Buffer.isBuffer(data)) {
    if (data instanceof ArrayBuffer) {
      data = Buffer.from(data);
    } else if (ArrayBuffer.isView(data)) {
      data = viewToBuffer(data);
    } else {
      data = Buffer.from(data);
      readOnly = false;
    }
  }

  data[kReadOnly] = readOnly;
  return data;
}
```

but I'm still not happy with it as that changes the buffer hidden class.

**Comment by lpinca** [comment]
I would like to have a LGTM from at least one of the other @websockets/admin members before merging.

**Comment by lpinca** [comment]
@alexions thanks but I'm -1 on that approach as that proposes to check the things twice (2x `Buffer.isBuffer()`, 2x `ArrayBuffer.isView()`, etc.), first to see if it's safe to treat the data as `readOnly` then to convert it to a buffer. I would rather the duplicated code.

**Comment by alexions** [comment]
ok. Agree that your changes will work faster.

**Comment by lpinca** [comment]
Before this change the whole `if...else`chain was used in 2 functions, now in 3 so I think it's not so bad. I'm just waiting for the approval of another @websockets/admin member :)

**Comment by lpinca** [comment]
I'll merge this tomorrow if there are no objections.

**Comment by lpinca** [comment]
@alexions thanks!

**Comment by alexions** [comment]
@lpinca you're welcome!

**Pull request merged**